### ISSUE TITLE
Mocha Setup is not working in typescript based tests.

### ISSUE THREAD

**Issue by raghanag** [Netflix/pollyjs#331]
## Description
I have setup mocha as defined in the setupMocha example. But when I run in replay mode, the recordings are created again thus having duplicate recordings. Please note that I am using ts-mocha to run the tests like this, and the below code is in one of the tests I have. 

### Shareable Source

this is how we run the tests.
    ```"test:integ": "ts-mocha -p ./tests/integ/tsconfig.json --opts ./tests/integ/mocha.opts",```
mocha.opts
```
--timeout=60000
--reporter mocha-multi-reporters
--reporter-options configFile=./tests/integ/report-config.json
./tests/integ/config/test-setup.ts
./tests/integ/**/*.ts


```

```js
const NodeHttpAdapter = require('@pollyjs/adapter-node-http');

const { Polly, setupMocha: setupPolly } = require('@pollyjs/core');
const FSPersister = require('@pollyjs/persister-fs');
const path = require('path');
Polly.register(NodeHttpAdapter);
Polly.register(FSPersister);
setupPolly({
    /* default configuration options */
    recordIfMissing: true,
    mode: 'replay',
    persister: 'fs',
    adapters: ['node-http'],
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../recordings')
      }
    }
  });
```

### Error Message & Stack Trace

```
There is no error but if I stop my app and run my functional tests then the tests should use the already recorded recordings since I set the mode as replay and also it should not record again and populate the recordings more.
```

### Dependencies

Copy the @pollyjs dependencies from `package.json`:

```json
    "@pollyjs/adapter": "^4.2.1",
    "@pollyjs/adapter-fetch": "^4.2.1",
    "@pollyjs/adapter-node-http": "^4.2.1",
    "@pollyjs/adapter-xhr": "^4.2.1",
    "@pollyjs/core": "^4.2.1",
    "@pollyjs/persister-fs": "^4.2.1",
    "@pollyjs/persister-rest": "^4.2.1",
```
## Environment

Node.js v10.15.0 darwin 17.7.0
npm -- 6.11.3

**Comment by raghanag** [comment]
@offirgolan I am tagging you to get your attention. Sorry about this.

**Comment by jasonmit** [comment]
A reproduction of the issue will speed up our turn around - unsure we'll be able to reproduce this with the given information,

**Comment by raghanag** [comment]
HI, @jasonmit. Sorry if I didn't give you enough info, here is the reproducible use case, I don't think its valid use case since its the selenium that is starting the chrome and launching the URL, I am more of looking for for example my test is launching google.com and just asserting what I key in the search value if you disconnect from the internet and run the test again in replay mode, it should pass, and also, for example, take any front-end URL and test it and record it and stop the back end services that are powering up the front-end test it replay mode, then it should pass. here is the repo
https://github.com/raghanag/polly-test. I will try to add a front-end and back-end URL instead of google.com

**Comment by raghanag** [comment]
And I looked at this issue, https://github.com/Netflix/pollyjs/issues/209 here the author said any requests made using selenium are recorded, but didn't give mode as record/replay in his example. I am not sure we both are looking for same issue.

**Comment by jasonmit** [comment]
Took a look at it, this configuration should resolve your issue:

```js
setupPolly({
  persister: 'fs',
  adapters: ['node-http'],
  persisterOptions: {
    fs: {
      recordingsDir: path.resolve(__dirname, '../recordings')
    }
  },
  matchRequestsBy: {
    headers(headers:Record<string, string>) {
      // `host` is unstable since the port changes between every run
      // probably configurable with Selenium to make it stable?
      delete headers.host;
      
      return headers;
    },
    url: {
      // same reason as above
      port: false,
    }
  }
});
```

**Comment by raghanag** [comment]
Hi I used the above configuration, without internet as i already have recordings, but the test is not getting passed, and the selenium is still launching the browser and trying to load google.com. And I see that mode and recordIfMissing options are missing in the above configuration. this is my config
```
setupPolly({
    /* default configuration options */
    recordIfMissing: true,
    mode: 'replay',
    persister: 'fs',
    adapters: ['node-http'],
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../recordings')
      }
    },
    matchRequestsBy: {
      headers(headers:Record<string, string>) {
        // `host` is unstable since the port changes between every run
        // probably configurable with Selenium to make it stable?
        delete headers.host;
        
        return headers;
      },
      url: {
        // same reason as above
        port: false,
      }
    }
  });
```

**Comment by jasonmit** [comment]
> Hi I used the above configuration, without internet as i already have recordings, but the test is not getting passed, and the selenium is still launching the browser and trying to load google.co

Polly is behaving how it's intended here.  Polly records anything that happens over node-http, xhr, and fetch.  What it doesn't record are all other requests (i.e., requests to load the document).

**Comment by raghanag** [comment]
I think i didn't clearly mention the problem I am facing, for example, if I test a front-end app using selenium/mocha/typescript (irrespective, just mentioning my tech stack) and the app using any back-end API's then I record the tests using pollyjs then even if the back-end API is down then the tests should pass in the replay mode. I hope I made myself clear and I will plan to add to my repo with the use case. I see that pollyjs only record the node-http, xhr and fetch and since the back-end calls from the front-end app will be done using by xhr mostly, then it should pass, but I tried the same in my app where I shut down my back-end services when I run in replay mode, the tests are getting failed.

**Comment by jasonmit** [comment]
> I record the tests using pollyjs then even if the back-end API is down then the tests should pass in the replay mode

Correct, that should be the case unless something is failing to match or wasn't recorded.

> I hope I made myself clear and I will plan to add to my repo with the use case.

Okay, look forward to seeing a reproduction - let me know once you push it to your repo and I'll have a look.

**Comment by raghanag** [comment]
I updated my app with replay mode and tested against the https://github.com/esausilva/example-create-react-app-express it has both front-end and back-end services, once I record the test, and replay the test when the back-end is down, it is working. Looks like PollyJS only records fetch, XHR. maybe my app is not using one of the methods. Thanks a lot @jasonmit

**Comment by jasonmit** [comment]
No problem!  I'm going to go ahead and close this.  Thanks for keeping us updated and putting together the example.

**Comment by raghanag** [comment]
Hi, @jasonmit Even though my configuration does not involve fetch, XHR other than node-http adapter in the setupPolly which is a setupMocha, may I know how is PollyJS recording the fetch and XHR API requests made by my front-end app. If you look at the configuration above, you can see I only added ```adapters: ['node-http']```

**Comment by jasonmit** [comment]
I'm not sure I understand the question @raghanag, can you rephrase?

**Comment by raghanag** [comment]
If you see at the line, the fetch is being used but I didn't mention it in my configuration above, I am trying to understand how the tests are being passed in replay mode. https://github.com/raghanag/example-create-react-app-express/blob/57ab2f2356cc659e782c75ab4613023fdf037a8c/client/src/App.js#L20

**Comment by raghanag** [comment]
And also when I changed from node-http to xhr in my mocha config, I see the below error.

Config
```
// const NodeHttpAdapter = require('@pollyjs/adapter-node-http');
import XHRAdapter from '@pollyjs/adapter-xhr';

const { Polly, setupMocha: setupPolly } = require('@pollyjs/core');
const FSPersister = require('@pollyjs/persister-fs');
const path = require('path');
Polly.register(XHRAdapter);
Polly.register(FSPersister);
describe('Selenium Demo Test Suite', () => {
  setupPolly({
    /* default configuration options */
    recordIfMissing: true,
    mode: 'record',
    persister: 'fs',
    adapters: ['xhr'],
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../recordings')
      }
    },
    matchRequestsBy: {
      headers(headers:Record<string, string>) {
        // `host` is unstable since the port changes between every run
        // probably configurable with Selenium to make it stable?
        delete headers.host;
        
        return headers;
      },
      url: {
        // same reason as above
        port: false,
      }
    }
  });
```

error
```
/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40
        var result = this._c[KEY](a === 0 ? 0 : a, b);
                                 ^
TypeError: Invalid value used as weak map key
    at WeakMap.set (<anonymous>)
    at WeakMap.set (/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40:34)
    at Function.register (/Users/naragha/samples/polly-test/node_modules/@pollyjs/core/src/polly.js:124:28)
    at Object.<anonymous> (/Users/naragha/samples/polly-test/tests/test.ts:11:7)
```

**Comment by raghanag** [comment]
When I used the below import into my test.ts file, I see the below error and looked at the https://github.com/Netflix/pollyjs/issues/98 and changed the import code to require. Now getting a different errors.
code
```
import XHRAdapter from '@pollyjs/adapter-xhr';
```
error
```
/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40
        var result = this._c[KEY](a === 0 ? 0 : a, b);
                                 ^
TypeError: Invalid value used as weak map key
    at WeakMap.set (<anonymous>)
    at WeakMap.set (/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40:34)
    at Function.register (/Users/naragha/samples/polly-test/node_modules/@pollyjs/core/src/polly.js:124:28)
    at Object.<anonymous> (/Users/naragha/samples/polly-test/tests/test.ts:11:7)
```

code
```
const XHRAdapter = require('@pollyjs/adapter-hxr');
```
error
```
1) Selenium Demo Test Suite
       "before each" hook for "should search for hello at localhost":
     PollyError: [Polly] [adapter:xhr] XHR global not found.
      at Object.assert (node_modules/@pollyjs/utils/src/utils/assert.js:5:11)
```

reading about https://github.com/Netflix/pollyjs/issues/199 but could not understand, can you explain little more in detail?

**Comment by raghanag** [comment]
HI, @jasonmit good morning. tagging you to get this issue to your attention. Sorry about this.

**Comment by jasonmit** [comment]
> PollyError: [Polly] [adapter:xhr] XHR global not found.

You're running the XHR adapter in Node and XHR API does not exist on Node.  You have two runtimes going: browser which selenium is opening up and Node which is running selenium webdriver.

**Comment by raghanag** [comment]
So the only option is to use node-http since I am running selenium tests in node and may I know how the PollyJS is recording the fetch or xhr requests that my browser is making if I don't mention that as part of my setupMocha config.

**Comment by jasonmit** [comment]
> So the only option is to use node-http since I am running selenium tests in node and may I know how the PollyJS is recording the fetch or xhr requests that my browser is making if I don't mention that as part of my setupMocha config.

If you're set on Selenium, you'll want to inject Polly into your app **not your selenium runtime**.  i.e., drop setupMocha and ensure whatever is running your web app has Polly instantiated with the right adapters setup.

The selenium story isn't one we've focused on.  If you want a similar testing experience, without the webdriver overhead, we recommend puppeteer instead.

**Comment by raghanag** [comment]
Can you please let me know how PollyJS is recording the calls made by the browser/front-end app to back-end app and recording if I didn't mention any fetch or xhr adapters in my setupMocha which is at selenium side.

**Comment by raghanag** [comment]
@jasonmit can you please clear my doubts, or please let me know if there is any slack channel/gitter room to post my questions there :)

**Comment by jasonmit** [comment]
> Can you please let me know how PollyJS is recording the calls made by the browser/front-end app to back-end app if I didn't mention any fetch or xhr adapters in my setupMocha which is at selenium side.

It's not - if you inspect the recording I don't think you'll see any of the requests that happened over fetch or xhr.  This is because those happened in a different context (the browser that selenium opens) and not your node process.

https://github.com/raghanag/polly-test/blob/master/recordings/Selenium-Demo-Test-Suite_3185844464/should-search-for-hello-at-localhost_2697116107/recording.har

**Comment by raghanag** [comment]
But the tests getting passed without the backend is up and running, if you see the har file line it has the back-end response recorded. Am I missing something? https://github.com/raghanag/polly-test/blob/master/recordings/Selenium-Demo-Test-Suite_3185844464/should-search-for-hello-at-localhost_2697116107/recording.har#L422

**Comment by jasonmit** [comment]
> But the tests getting passed without the backend is up and running

It's recording your node test interfacing with Selenium.  Not Selenium interfacing with your web app.  That response you linked me to is coming from the `p.getText();` call.

**Comment by raghanag** [comment]
> I believe this all can be achieved, I just don't know the Selenium API off the top of my head.
Can you let me know how to achieve it as the examples section or documentation does not have what I am looking for.

**Comment by jasonmit** [comment]
We have examples using Puppeteer instead of Selenium: https://github.com/Netflix/pollyjs/blob/master/examples/jest-puppeteer/__tests__/dummy-app.test.js

**Comment by raghanag** [comment]
Hi, @jasonmit I have tried to use puppeteer, puppeteer-adapter instead of selenium with mocha, typescript. But still seeing the below error. I have uploaded the changes to my repo. you can run the test against the dummy-app example in PollyJS examples. 
my repo: https://github.com/raghanag/polly-test
```
/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40
        var result = this._c[KEY](a === 0 ? 0 : a, b);
                                 ^
TypeError: Invalid value used as weak map key
    at WeakMap.set (<anonymous>)
    at WeakMap.set (/Users/naragha/samples/polly-test/node_modules/core-js/library/modules/_collection.js:40:34)
    at Function.register (/Users/naragha/samples/polly-test/node_modules/@pollyjs/core/src/polly.js:124:28)
    at Object.<anonymous> (/Users/naragha/samples/polly-test/tests/test.ts:11:7)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
```

**Comment by jasonmit** [comment]
@raghanag your environment isn't configured correctly.  Set `esModuleInterop`

![image](https://user-images.githubusercontent.com/3108309/82714167-59683e80-9c42-11ea-8cca-142eed487d86.png)

**Comment by raghanag** [comment]
Hi @jasonmit I changed it and now the tests are failing with the below error, all I did was to use puppeteer and copied the same setupPolly from your puppeteer example and used mocha instead of jest. just updated my repo

```
1) Puppeteer Suite
       should be able to navigate to all routes:
     TypeError: Cannot read property 'polly' of undefined
      at /Users/naragha/samples/polly-test/tests/test.ts:68:21
      at step (node_modules/tslib/tslib.js:140:27)
      at Object.next (node_modules/tslib/tslib.js:121:57)
      at fulfilled (node_modules/tslib/tslib.js:111:62)
      at process._tickCallback (internal/process/next_tick.js:68:7)

  2) "after each" hook for "should be able to navigate to all routes":
     TypeError: Cannot read property 'stop' of undefined
      at afterEach (node_modules/@pollyjs/core/src/test-helpers/lib.js:17:23)
```

**Comment by jasonmit** [comment]
You're using setupMocha wrong:

https://github.com/Netflix/pollyjs/blob/master/docs/test-frameworks/mocha.md#simple-example-docsify-ignore

**Comment by raghanag** [comment]
Is this the right way to use https://github.com/raghanag/polly-test/blob/master/tests/test.ts

I am still not getting the Polly instance to assign puppeteer in my test(it).
```
TypeError: Cannot read property 'configure' of undefined
      at /Users/naragha/samples/polly-test/tests/test.ts:28:13
```

**Comment by jasonmit** [comment]
No it's not correct, please look at the example closer.  `this.polly` and you would therefore also need to change `async () => ` to `async function() {` so that you have `this` scoped correctly.

I cannot continue to circle back to this, we have great examples and documentation around this already.

**Comment by raghanag** [comment]
Sorry @jasonmit , I have fixed my example and the back-end call is now getting recorded. It looks like we have to strictly follow the documentation. https://github.com/raghanag/polly-test/blob/master/recordings/Puppeteer-Suite_952821357/should-be-able-to-navigate-to-all-routes_1130491217/recording.har#L54
